control/control.go

```diff
@@ -35,51 +32,42 @@ func NewHTTPServer(cfg *Config) (*HTTPServer, error) {
 	}
 
 	mux := http.NewServeMux()
-	addr := net.UnixAddr{
-		Name: cfg.SocketPath,
-		Net: SocketType,
-	}
 
 	return &HTTPServer{
+		Addr: cfg.SocketPath,
 		mux: mux,
-		addr: addr,
-		listening: false,
 	}, nil
 }
 
-var listener net.Listener
-
 // Serve starts serving the control server
-func (s *HTTPServer) Serve() {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	// ref https://github.com/joyent/containerpilot/pull/165
-	if listener != nil {
-		return
-	}
+func (s *HTTPServer) Start(app interface{}) {
+	// s.lock.Lock()
+	// defer s.lock.Unlock()
 
 	s.mux.HandleFunc("/v3/env", s.getEnvHandler)
+	s.mux.HandleFunc("/v3/reload", s.postReloadHandler)
+
+	s.Handler = s.mux
 
-	ln, err := net.Listen(s.addr.Network(), s.addr.String())
+	ln, err := net.Listen(SocketType, s.Addr)
 	if err != nil {
-		log.Fatalf("error serving socket at %s: %v", s.addr.String(), err)
+		log.Fatalf("error serving socket at %s: %v", s.Addr, err)
 	}
 
-	listener = ln
-	s.listening = true
-
 	go func() {
-		log.Infof("control: Serving at %s", s.addr.String())
-		log.Fatal(http.Serve(ln, s.mux))
-		log.Debugf("control: Stopped serving at %s", s.addr.String())
+		log.Infof("control: Serving at %s", s.Addr)
+		log.Fatal(s.Serve(ln))
+		log.Debugf("control: Stopped serving at %s", s.Addr)
 	}()
 }
 
-// Shutdown shuts down the control server
-func (s *HTTPServer) Shutdown() {
-	s.listening = false
-	log.Debug("control: Shutdown received but currently a no-op")
+// Stop shuts down the control server gracefully
+func (s *HTTPServer) Stop() {
+	if err := s.Shutdown(nil); err != nil {
+		log.Fatal("control: failed to shutdown HTTP control plane")
```

Comment:
First go around I wasn't considering the configuration reload here so it was definitely drastic. I'll return the error back to the app reload context so it can be handled accordingly there.